trick_sims/SIM_singlerigidbody/models/src/body.cpp

```diff
@@ -0,0 +1,353 @@
+	
+
+	for(int i = 0; i<6; i++)
+		vec_b[i] = 0.0;
+
+	
+	for(int i = 0; i<6; i++)
+		for(int j = 0; j<6; j++)
+			mat_mass[i][j] = 0.0;
+	V_INIT(pos);
+	V_INIT(vel);
+	V_INIT(acc);
+	V_INIT(force);
+	V_INIT(force_remain);
+	V_INIT(torque);
+	V_INIT(torque_remain);
+	V_INIT(angle);
+	V_INIT(omega);
+	V_INIT(omegaDot);
+	V_INIT(angle_force);
+
+	M_INIT(CM_skew);
+	M_INIT(m_CM_skew);
+	M_INIT(massmatrix);
+	M_INIT(rotate);
+
+
+	// Inputs 
+	mass = 20.0;
+	radius = 2.0;
+
+	POS_INIT[0] = 0.0;
+	POS_INIT[1] = 0.0;
+	POS_INIT[2] = 0.0;
+
+	ANGLE_INIT[0] = 0.0;
+	ANGLE_INIT[1] = 0.0;
+	ANGLE_INIT[2] = 0.0;
+
+	OMEGA_INIT[0] = 0.0;
+	OMEGA_INIT[1] = 0.0;
+	OMEGA_INIT[2] = 0.0;
+
+	FORCE_INIT[0] = 0.0;
+	FORCE_INIT[1] = 0.0;
+	FORCE_INIT[2] = 10.0;
+
+	VEL_INIT[0] = 0.0;
+	VEL_INIT[1] = 0.0;
+	VEL_INIT[2] = 0.0;
+
+	POS_FORCE[0] = 2.0;
+	POS_FORCE[1] = 0.0;
+	POS_FORCE[2] = 0.0;
+
+};
+
+/************************************************************************
+FUNCTION: SingleRigidBody::init()
+PURPOSE: (Initializing state variables, creating inertia and CoM skew 
+	  matrices) 
+************************************************************************/
+
+void BODY::init() {
+
+	V_COPY(pos, POS_INIT);
+	V_COPY(vel, VEL_INIT);
+	V_COPY(angle, ANGLE_INIT);	
+	V_COPY(omega, OMEGA_INIT);
+
+	// Center of mass skew matrix 	
+	V_INIT(CM);
+
+	V_SKEW(CM_skew, CM);
+
+	MxSCALAR(m_CM_skew, CM_skew, mass);	
+	MxSCALAR(neg_m_CM_skew, m_CM_skew, -1.0);
+
+	// Inertia matrix
+	inertia = (2.0/5.0) * mass * radius * radius; 
+
+	for(int i = 0; i<1; i++)
+                for(int j = 0; j<1; j++)
+                        inertia_matrix[i][j] = inertia;
+
+        for(int i = 1; i<2; i++)
+                for(int j = 1; j<2; j++)
+                        inertia_matrix[i][j] = inertia;
+
+        for(int i = 2; i<3; i++)
+                for(int j = 2; j<3; j++)
+                        inertia_matrix[i][j] = inertia;
+
+};
+
+/******************************************************************************
+FUNCTION: BODY::mass_matrix()
+PURPOSE: (Calculates 6x6 mass matrix M for eq M*a + r = b)
+******************************************************************************/
+void BODY::mass_matrix(){
+
+	// Creates mass matrix for mat_mass()	
+	for(int i = 0; i<1; i++)
+		for(int j = 0; j<1; j++)
+			massmatrix[i][j] = mass; 
+
+	for(int i = 1; i<2; i++)
+		for(int j = 1; j<2; j++)
+			massmatrix[i][j] = mass; 
+
+	for(int i = 2; i<3; i++)
+		for(int j = 2; j<3; j++)
+			massmatrix[i][j] = mass; 
+	
+
+	// Combines inertia, center of mass, and mass matrix into a 6x6 matrix 
+	for(int i = 0; i<3; i++)
+		for(int j = 0; j<3; j++)
+			mat_mass[i][j] = massmatrix[i][j];
+
+	for(int i = 0; i<3; i++)
+		for(int j = 3; j<6; j++)
+			mat_mass[i][j] = neg_m_CM_skew[i][j-3];
+
+	for(int i = 3; i<6; i++)
+		for(int j = 0; j<3; j++)
+			mat_mass[i][j] = m_CM_skew[i-3][j];
+
+	for(int i = 3; i<6; i++)
+		for(int j = 3; j<6; j++)
+			mat_mass[i][j] = inertia_matrix[i-3][j-3];
+
+};
```

Comment:
If all of these values are constant, is there a reason that this routine is called in the derivative loop instead of at init?  There's no reason, if the mass matrix is constant, to continually reassign and recalculate the matrix.  It can just be done at init.